File: maximum-binary-tree.cpp
//Runtime: 62 ms
class Solution {
public:

    TreeNode* solve(vector<int>&nums, int x, int y)
    {
        if (x > y)
            return NULL;
        int maxn = INT_MIN;
        int maxi = -1;
        for (int i = x; i <= y; i++)
        {
            if (maxn < nums[i])
            {
                maxn = nums[i];
                maxi = i;
            }
        }

        TreeNode* node = new TreeNode(nums[maxi]);
        node->left = solve(nums, x, maxi - 1);
        node->right = solve(nums, maxi + 1, y);
        return node;
    }

    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        return solve(nums, 0, nums.size() - 1);
    }
};

File: maximum-binary-tree.py
# Runtime: 239 ms
class Solution(object):
    def constructMaximumBinaryTree(self, nums):
        if len(nums) == 0:
            return None

        t = 0
        for i in range(len(nums)):
            if nums[t] < nums[i]:
                t = i

        root = TreeNode(nums[t])
        root.left = self.constructMaximumBinaryTree(nums[:t])
        root.right = self.constructMaximumBinaryTree(nums[t + 1:])
        return root
